File: DotNet Project V2.0/backend-dotnet/src/Emart.Api/Controllers/SubCategoryController.cs
using Emart.Domain.Entities;
using Emart.Application.Services.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Cors;
using Microsoft.AspNetCore.Mvc;

namespace Emart.Api.Controllers;

// Same authorization shape as CategoryController: GET falls through to
// anyRequest().authenticated() in the source SecurityConfig - no anonymous browsing.
[ApiController]
[Route("api/subcategory")]
[EnableCors("AnyOrigin")]
[Authorize]
public class SubCategoryController : ControllerBase
{
    private readonly ISubCategoryService _service;

    public SubCategoryController(ISubCategoryService service)
    {
        _service = service;
    }

    [HttpPost]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<SubCategory>> Save([FromBody] SubCategory subCategory, CancellationToken ct) =>
        Ok(await _service.SaveAsync(subCategory, ct));

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<SubCategory>>> GetAll(CancellationToken ct) =>
        Ok(await _service.GetAllAsync(ct));

    [HttpGet("{id:int}")]
    public async Task<ActionResult<SubCategory>> GetById(int id, CancellationToken ct) =>
        Ok(await _service.GetByIdAsync(id, ct));

    [HttpGet("category/{catmasterId:int}")]
    public async Task<ActionResult<IReadOnlyList<SubCategory>>> GetByCategoryId(int catmasterId, CancellationToken ct) =>
        Ok(await _service.GetByCategoryIdAsync(catmasterId, ct));

    [HttpPut("{id:int}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<SubCategory>> Update(int id, [FromBody] SubCategory subCategory, CancellationToken ct) =>
        Ok(await _service.UpdateAsync(id, subCategory, ct));

    [HttpDelete("{id:int}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<string>> Delete(int id, CancellationToken ct)
    {
        await _service.DeleteAsync(id, ct);
        return Ok("SubCategory Deleted Successfully");
    }
}

File: Java Project  V1.7/backend-spring/src/main/java/com/emart/controller/SubCategoryController.java
package com.emart.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import com.emart.entity.SubCategory;
import com.emart.service.SubCategoryService;

@RestController
@RequestMapping("/api/subcategory")
@CrossOrigin("*")
public class SubCategoryController {

    @Autowired
    private SubCategoryService service;


    @PostMapping
    public SubCategory save(
            @RequestBody SubCategory subCategory) {

        return service.save(subCategory);
    }


    @GetMapping
    public List<SubCategory> getAll() {

        return service.getAll();
    }


    @GetMapping("/{id}")
    public SubCategory getById(
            @PathVariable Integer id) {

        return service.getById(id);
    }


    @GetMapping("/category/{catmasterId}")
    public List<SubCategory> getByCategoryId(
            @PathVariable Integer catmasterId) {

        return service.getByCategoryId(
                catmasterId
        );
    }


    @PutMapping("/{id}")
    public SubCategory update(
            @PathVariable Integer id,
            @RequestBody SubCategory subCategory) {

        return service.update(
                id,
                subCategory
        );
    }


    @DeleteMapping("/{id}")
    public String delete(
            @PathVariable Integer id) {

        service.delete(id);

        return "SubCategory Deleted Successfully";
    }
}